### ISSUE TITLE
Figure 2 in Paper

### ISSUE THREAD

**Issue by lazyeden** [aoi3142/Waterfall#1]
Could you explain more about the meaning of Figure 2 in the paper, especially Figure 2c, such as why we need calculate the mean of the permutations, and why L + G shows less distortion to G in Vw space, and to L in Vo space.

**Comment by aoi3142** [comment]
Hi @redwyd, thanks for checking out Waterfall!

For each token that is being generated from the watermarking process, the watermarked logits $\check{L}$ contain both the original logits $L$ that convey the semantic meaning of the text, and the watermarking function $G$. However, the original logits $L$ itself has some signal (which is dissimilar to the watermarking function), which interferes with the watermarking function $G$, affecting the detection of the watermark in the generated text (verifiability). The same is true where the watermarking function $G$ affects the original logits $L$, affecting the semantic content of the generated text (fidelity).

However, with the use of the pseudo-random permutation function $P(k_\pi, L)$ during watermarking, for each token that is generated, the randomly permuted logit distribution $L$ (for the different tokens across the text) "flattened out" into a flat distribution $\bar{L}$ when viewing in the permuted $V_w$ space, so that the watermarking function $G$ "stands out" from the flat $\bar{L}$ "noise", making it easy to verify. This similarly applies to the logits signal when viewed in original $V_o$ space, where the original logits $L$ will stand out from the "flattened out" $\bar{G}$ due to the random permutations of the watermark function $G$, making them more aligned to the actual generation probability of the original LLM, across the entire text.

Although Figure 2 is an idealized case for a toy example, we can observe this behaviour of the random permutations "flattening out" empirically. Figure 8 shows the watermarking signal we are interested in looking for "standing out" in real watermarked text. In addition, Figure 9 shows that the signal only appears when using the correct watermarking key $k_\pi$ to get the correct pseudo-random permutations for the $V_w$ space, while watermarked text verified with the wrong key will not show the signal. We also provide further experimental results showing the fidelity and verifiability of Waterfall in Figure 4 and 5.

I hope this answers your question, do reach out if you have further queries! Thanks!

**Comment by lazyeden** [comment]
Thank you for your answer, but I am still a little confused. When generating each token, the permutation function is only used once, so why do I need to get a flat distribution $L$ when viewing in the permuted $V_w$ space? Similarly, when generating each token, only one watermark signal is added, so why do I need to calculate a flat distribution $G$ when viewing in the permuted $V_0$ space? In addition, why can't we operate directly in the original space, but have to permute it and then permute it back to the original space?

**Comment by aoi3142** [comment]
Thanks @redwyd for following up!

To clarify, there is a single $V_o$ space for all tokens in the watermarked text (the token ordering in the model's tokenizer), but each token generated has its own different $V_w$ space. The index permutation in each $V_w$ space is determined by the watermark key and preceeding $n$ tokens (Figure 3). As the preceeding $n$ tokens vary across the text, each token has a different random $V_w$ space permutation, where each possible permutation is equally likely to be used (large green arrow in Figure 2c) due to the pseudo-random permutation.

You are right in saying that the logits distribution $L$ does not flatten out when generating each single token. However, as we are sampling more and more tokens as the watermarked text is generated, each time round, the logits are permuted with a different pseudo-random permutation to its own $V_w$ space. After many different random permutations across the different tokens generated, these different permutations of $L$ in across the different $V_w$ space flatten out when viewing all the $V_w$ spaces together. More specifically, in Figure 2, the distribution $\bar{L}$ flattens out only after averaging many different permutations $P(1,L),\ldots,P(6,L)$ across many generations. This flattening out is needed so than when considering the watermarked text as a whole, the watermark signal emerges from the flattened $\bar{L}$ (verifiability desideratum).

The purpose for apply the watermark function $G$ after the permutation to the $V_w$ space is to make sure that $G$ does not adversely affect the underlying distribution of the watermarking LLM. If no permutation is done, those tokens that are at the indices where $G$ is high (e.g., peaks in the sine function) will have boosted generation probablity across the entire text, causing them to be generated with higher probabilities everywhere in the text, no matter whether they are relavant or not. By applying $G$ in the permuted $V_w$ space, different tokens will be randomly boosted at different locations across the text, minimizing the overall impact on the watermarked text (fidelity desideratum).

Furthermore, watermarking only in $V_o$ space limits the number of data owners that Waterfall can support (scalability desideratum). If all users watermark in the same $V_o$ space, we cannot tell different data owners apart. Additionally, this also makes it easier for any potential adversaries to make guesses about $G$ by observing high frequency tokens that seem out of place in the watermarked text, thereby stealing the watermark, erasing the watermark, or impersonate the data owner by creating new watermarked text that will be falsely attributed to the data owner, even without knowledge of the true $G$ or watermarking key (robust verifiability desideratum).

By using a seeded pseudo-random permutation, the signal applied by the $G$ is not directly observable in $V_o$ space, and only appears when permuting the token distribution into $V_w$ space with the correct key.

P.S. The actual implementation of the code in this GitHub repo does not actually permute the logits twice ($L$ into $V_w$ space, adding $G$, then permuting it back into $V_o$ space). Instead, only 1 permutation operation is used for efficiency, where $G$ is permuted from $V_w$ space into $V_o$ space, and added to $L$ in $V_o$ space.

**Comment by lazyeden** [comment]
Thank you for your patience, I think I understand.

**Comment by aoi3142** [comment]
No problem! Feel free to reach out again if you have any other questions!
